core/tauri/src/tray/plugin.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright 2019-2023 Tauri Programme within The Commons Conservancy
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-License-Identifier: MIT
+
+use std::path::PathBuf;
+
+use serde::Deserialize;
+
+use crate::{
+  command,
+  menu::{plugin::ItemKind, Menu, Submenu},
+  plugin::{Builder, TauriPlugin},
+  resources::ResourceId,
+  tray::TrayIconBuilder,
+  AppHandle, IconDto, Manager, Runtime,
+};
+
+use super::TrayIcon;
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct TrayIconOptions {
+  id: Option<String>,
+  menu: Option<(ResourceId, ItemKind)>,
+  icon: Option<IconDto>,
+  tooltip: Option<String>,
+  title: Option<String>,
+  temp_dir_path: Option<PathBuf>,
+  icon_as_template: Option<bool>,
+  menu_on_left_click: Option<bool>,
+}
+
+#[command(root = "crate")]
+fn new<R: Runtime>(
+  app: AppHandle<R>,
+  options: TrayIconOptions,
+) -> crate::Result<(ResourceId, String)> {
+  let mut builder = if let Some(id) = options.id {
+    TrayIconBuilder::<R>::with_id(id)
+  } else {
+    TrayIconBuilder::<R>::new()
+  };
+
+  builder = builder.on_tray_event(|tray, e| {
+    let _ = tray.app_handle().emit_all("tauri://tray", e);
```

Comment:
I think we should use a channel here instead, so we know only the tray icon caller have access to these events

also simplifies the JS APi a lot :)